pennylane/qnode_new/jacobian.py

```diff
@@ -0,0 +1,528 @@
+        mutable = self.mutable
+        self.mutable = False
+
+        if wrt is None:
+            wrt = range(self.num_variables)
+        else:
+            if min(wrt) < 0 or max(wrt) >= self.num_variables:
+                raise ValueError("Tried to compute the gradient wrt. free parameters {} "
+                                 "(this node has {} free parameters).".format(wrt, self.num_variables))
+            if len(wrt) != len(set(wrt)):  # set removes duplicates
+                raise ValueError("Parameter indices must be unique.")
+
+        # check if the method can be used on the requested parameters
+        method_map = _inv_dict(self.par_to_grad_method)
+        def inds_using(m):
+            """Intersection of ``wrt`` with free params indices whose best grad method is m."""
+            return method_map.get(m, set()).intersection(wrt)
+
+        # are we trying to differentiate wrt. params that don't support any method?
+        bad = inds_using(None)
+        if bad:
+            raise ValueError('Cannot differentiate wrt. parameters {}.'.format(bad))
+
+        if method in ('A', 'F'):
+            if method == 'A':
+                bad = inds_using('F')
+                if bad:
+                    raise ValueError("The analytic gradient method cannot be "
+                                     "used with the parameters {}.".format(bad))
+            method = {k: method for k in wrt}
+        elif method == 'B':
+            # use best known method for each parameter
+            method = self.par_to_grad_method
+        else:
+            raise ValueError('Unknown gradient method.')
+
+        if 'F' in method.values():
+            if options.get('order', 1) == 1:
+                # the value of the circuit at args, computed only once here
+                options['y0'] = np.asarray(self.evaluate(args, kwargs))
+
+        # flatten the nested Sequence of input arguments
+        flat_args = np.array(list(_flatten(args)), dtype=float)
+        variances_required = any(ob.return_type is ObservableReturnTypes.Variance for ob in self.circuit.observables)
+
+        # compute the partial derivative wrt. each parameter using the appropriate method
+        grad = np.zeros((self.output_dim, len(wrt)), dtype=float)
+        for i, k in enumerate(wrt):
+            if k not in self.variable_deps:
+                # unused parameter, partial derivatives wrt. it are zero
+                continue
+
+            par_method = method[k]
+            if par_method == 'A':
+                if variances_required:
+                    grad[:, i] = self._pd_parameter_shift_var(k, flat_args, kwargs)
+                else:
+                    grad[:, i] = self._pd_parameter_shift(k, flat_args, kwargs, **options)
+            elif par_method == 'F':
+                grad[:, i] = self._pd_finite_diff(k, flat_args, kwargs, **options)
+            else:
+                raise ValueError('Unknown gradient method.')
```

Comment:
what about the new `D` method? It might be expected from the docstring that you can use `D`, but this error message would be contradictory

Reply:
As far as I can tell, this doesn't seem the best place to have the custom gradient option, as it wouldn't be a per-parameter option, right? Instead, if the device supports gradients, it should use a QNode that overwrites this entire method with it's own jacobian function.

Reply:
The 'D' method should just call `self.device.jacobian` and return whatever it gives. I'll fix it asap.